src/osdc/Objecter.h

```diff
@@ -431,19 +435,28 @@ struct ObjectOperation {
   struct CB_ObjectOperation_cmpext {
     int* prval = nullptr;
     boost::system::error_code* ec = nullptr;
-    std::size_t* s = nullptr;
+    uint64_t* s = nullptr;
     explicit CB_ObjectOperation_cmpext(int *prval)
       : prval(prval) {}
-    CB_ObjectOperation_cmpext(boost::system::error_code* ec, std::size_t* s)
+    CB_ObjectOperation_cmpext(boost::system::error_code* ec, uint64_t* s)
       : ec(ec), s(s) {}
 
-    void operator()(boost::system::error_code ec, int r, const ceph::buffer::list&) {
+    void operator()(boost::system::error_code ec, int r,
+		    const ceph::buffer::list&) {
       if (prval)
         *prval = r;
-      if (this->ec)
-	*this->ec = ec;
+      if (this->ec) {
+	if (r < 0) {
+	  *this->ec = make_error_code(osd_errc::cmpext_failed);
+	} else {
+	  this->ec->clear();
+	}
+      }
       if (s)
-	*s = static_cast<std::size_t>(-(MAX_ERRNO - r));
+	*s = r == 0 ? -1 : -MAX_ERRNO - r;
+      if (r < 0 && r < -MAX_ERRNO) {
+	throw boost::system::system_error(osd_errc::cmpext_failed);
```

Comment:
Objecter did not have a way of letting the handler add or change the error code. I added one in at Casey's suggestion to handle parse errors in CLS handlers. It uses exceptions because that lets all the handlers and CLS clients that currently exist work without having to be modified to use a different argument signature.

Reply:
Or, more specifically, it can't amend the error of the OSDOp as a whole. You can pass to a supplied pointer. But this makes the error handling much more awkward for the client.

Reply:
this feels different than the cls case where the response parsing errors only happen on the client side. if CEPH_OSD_OP_CMPEXT failed on the server-side, it should have filled in the osd op's error code itself

Reply:
The OSD uses its single integer return value to specify the offset failure as a number less than `-ERRNO_MAX`. This does not work with the `error_code` library which needs a *specific* code, which we're supplying in the client handler.

Reply:
> The OSD uses its single integer return value to specify the offset failure as a number less than -ERRNO_MAX. This does not work with the error_code library which needs a specific code, which we're supplying in the client handler.

you could potentially make `cmpext_failed` an error condition in [ceph_category](https://github.com/ceph/ceph/blob/f64ad440/src/common/error_code.h#L37-L48) and use [default_error_condition()](https://en.cppreference.com/w/cpp/error/error_category/default_error_condition) to map values below `-ERRNO_MAX` to it
```c++
if (ec == ceph::errc::cmpext_failed) {
  // convert ec.value() back to an offset
}
```
it just wouldn't make sense to do something like `make_error_code(cmpext_failed)`

Reply:
I could, but that sounds much worse. I don't see an advantage to doing it that way. Since we have the machinery to add/change an error-code in the client and we need a specific error code, why not use it?